// @loaders.gl/i3s/dist/es5/lib/parsers/constants.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FLOAT_64_TYPE = exports.OBJECT_ID_ATTRIBUTE_TYPE = exports.STRING_ATTRIBUTE_TYPE = exports.SIZEOF = exports.I3S_NAMED_HEADER_ATTRIBUTES = exports.I3S_NAMED_GEOMETRY_ATTRIBUTES = exports.I3S_NAMED_VERTEX_ATTRIBUTES = exports.GL_TYPE_MAP = exports.TYPE_ARRAY_MAP = void 0;

var _constants = _interopRequireDefault(require("@luma.gl/constants"));

var TYPE_ARRAY_MAP = {
  UInt8: Uint8Array,
  UInt16: Uint16Array,
  UInt32: Uint32Array,
  Float32: Float32Array,
  UInt64: Float64Array
};
exports.TYPE_ARRAY_MAP = TYPE_ARRAY_MAP;
var GL_TYPE_MAP = {
  UInt8: _constants.default.UNSIGNED_BYTE,
  UInt16: _constants.default.UNSIGNED_INT,
  Float32: _constants.default.FLOAT,
  UInt32: _constants.default.UNSIGNED_INT,
  UInt64: _constants.default.DOUBLE
};
exports.GL_TYPE_MAP = GL_TYPE_MAP;
var I3S_NAMED_VERTEX_ATTRIBUTES = {
  position: 'position',
  normal: 'normal',
  uv0: 'uv0',
  color: 'color',
  region: 'region'
};
exports.I3S_NAMED_VERTEX_ATTRIBUTES = I3S_NAMED_VERTEX_ATTRIBUTES;
var I3S_NAMED_GEOMETRY_ATTRIBUTES = {
  vertexAttributes: 'vertexAttributes',
  featureAttributeOrder: 'featureAttributeOrder',
  featureAttributes: 'featureAttributes'
};
exports.I3S_NAMED_GEOMETRY_ATTRIBUTES = I3S_NAMED_GEOMETRY_ATTRIBUTES;
var I3S_NAMED_HEADER_ATTRIBUTES = {
  header: 'header',
  vertexCount: 'vertexCount',
  featureCount: 'featureCount'
};
exports.I3S_NAMED_HEADER_ATTRIBUTES = I3S_NAMED_HEADER_ATTRIBUTES;
var SIZEOF = {
  UInt8: 1,
  UInt16: 2,
  UInt32: 4,
  Float32: 4,
  UInt64: 8
};
exports.SIZEOF = SIZEOF;
var STRING_ATTRIBUTE_TYPE = 'String';
exports.STRING_ATTRIBUTE_TYPE = STRING_ATTRIBUTE_TYPE;
var OBJECT_ID_ATTRIBUTE_TYPE = 'Oid32';
exports.OBJECT_ID_ATTRIBUTE_TYPE = OBJECT_ID_ATTRIBUTE_TYPE;
var FLOAT_64_TYPE = 'Float64';
exports.FLOAT_64_TYPE = FLOAT_64_TYPE;
