// @loaders.gl/i3s/dist/es5/i3s-node-page-loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.I3SNodePageLoader = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var VERSION = typeof "3.0.12" !== 'undefined' ? "3.0.12" : 'latest';

function parseNodePage(_x) {
  return _parseNodePage.apply(this, arguments);
}

function _parseNodePage() {
  _parseNodePage = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee(data) {
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt("return", JSON.parse(new TextDecoder().decode(data)));

          case 1:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _parseNodePage.apply(this, arguments);
}

var I3SNodePageLoader = {
  name: 'I3S Node Page',
  id: 'i3s-node-page',
  module: 'i3s',
  version: VERSION,
  mimeTypes: ['application/json'],
  parse: parse,
  extensions: ['json'],
  options: {}
};
exports.I3SNodePageLoader = I3SNodePageLoader;

function parse(_x2) {
  return _parse.apply(this, arguments);
}

function _parse() {
  _parse = (0, _asyncToGenerator2.default)(_regenerator.default.mark(function _callee2(data) {
    return _regenerator.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            data = parseNodePage(data);
            return _context2.abrupt("return", data);

          case 2:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _parse.apply(this, arguments);
}
